### ISSUE TITLE
[Breaking Change Request] [vm] Deprecate and remove dart native extensions

### ISSUE THREAD

**Issue by mraleph** [dart-lang/sdk#45451]
# Summary

Dart Native Extensions is a mechanism for loading native code into a running Dart program and binding Dart methods to native implementatations. 

It consists of the following parts: 

* `dart-ext:...` import scheme which instructs VM to load a native library;
* `native` keyword which specifies that a function has a native implementation;
* Dart VM C API (as defined in `runtime/include/dart_api.h`) which can be used to reflectively inspect and mutate the state of the Dart program;  

Currently native extensions only really work in a standalone Dart VM - Flutter embedder hides VM C API symbols and does not allow developers to interact with Dart through Dart VM APIs.

Relience on C API means that:

* Native extensions are not very efficient for fine-grained binding.
* Reflective access through Dart VM C API is also incompatible with AOT (requires careful usage of `@pragma('vm:entry-point')` which in turn disables treeshaking).  

`dart:ffi` was designed to not have these problems - be both more effecient and more AOT friendly. `dart:ffi` works in standalone VM as well as Flutter.

From our perspective it does not make sense to maintain two completely different approaches to binding Dart code to native especially given that one of them is obviously superior. We don't want more packages using native extensions to appear in the ecosystem and we want to push all new Dart users to use `dart:ffi`. 

We propose the following:

* In Dart 2.13 issue a deprecatin warning whenever analyzer or CFE see `dart-ext:` import.  
* In Dart 2.14 remove support for `dart-ext:` imports and hide Dart VM C API symbols from standalone `dart` binary shipped with SDK

# Impact

All packages that use `dart-ext:` will be broken by the release. We will work to estimate the impact, but we consider it to be very small because none of the Flutter compatible packages will be affected, as Flutter does not support this mechanism in the first place.

**Comment by a-siva** [comment]
Adding the March milestone for the deprecation warning in 2.13 from analyzer and CFE when dart-ext import is seen.

In 2.14 we should also remove `samples/sample_extension`

**Comment by a-siva** [comment]
Also added area-analyzer and area-front-end labels for the warning work in 2.13

**Comment by devoncarew** [comment]
A deprecation warning wouldn't be much work, but we probably want to discuss adding new work to a release a few working days before a branch cut.

**Comment by mraleph** [comment]
@devoncarew I have updated versions to 2.14 and 2.15 and moved the milestone. No intention to fall over ourselves with this.

**Comment by blagoev** [comment]
Does Dart SDK follow semver versioning?

**Comment by franklinyow** [comment]
Ping @vsmenon @matanlurey @Hixie for approval

**Comment by Hixie** [comment]
Fine by me since this doesn't affect Flutter anyway. :-)

**Comment by vsmenon** [comment]
lgtm

**Comment by franklinyow** [comment]
Approved

**Comment by franklinyow** [comment]
@mraleph Please add a post to https://groups.google.com/a/dartlang.org/g/announce

**Comment by mraleph** [comment]
@devoncarew @johnniwinther does it sound reasonable to have analyzer produce deprecation warnings in 2.14? I want to get versions right in the announcement.

**Comment by johnniwinther** [comment]
SGTM

Do you have wording to use in the deprecation warning?

**Comment by mraleph** [comment]
@johnniwinther I suggest "Dart native extensions are deprecated and will be removed in 2.15 release of Dart SDK. Migrate to using FFI instead (https://dart.dev/guides/libraries/c-interop)". @mit-mit any thoughts?

**Comment by srawlins** [comment]
What is the plan for moving internal code off  of dart-ext? I see a few cases when I search `lang:dart import\ ['"]dart-ext` in code search.

**Comment by mraleph** [comment]
I have reached to some of the internal owners and working on planning migration

**Comment by devoncarew** [comment]
> I have reached to some of the internal owners and working on planning migration

Great! I assume internal users would be able to ignore the deprecation message at the use sites, in which case we could do the analyzer work - the deprecation message - w/o being blocked on removing uses internally first.

**Comment by mraleph** [comment]
FWIW, currently I think there is actually only 1 user that needs to be migrated. Other uses are either on legacy Dart v1 or are can be deleted. I am planning to migrate that 1 user myself. I will update this thread once I have all the data.

**Comment by franklinyow** [comment]
@mraleph @devoncarew @johnniwinther 
Should we have the warning message for CFE and analyzer be release on the same month?

Currently CFE  is on current milestone and analyzer is on May

**Comment by mraleph** [comment]
I am fine with May milestone for both

**Comment by mit-mit** [comment]
Before we do this, I think we need a replacement for https://dart.dev/server/c-interop-native-extensions that describes (at least in some detail) how to do integration/embedding with FFI.

**Comment by devoncarew** [comment]
> Should we have the warning message for CFE and analyzer be release on the same month?
> Currently CFE is on current milestone and analyzer is on May

Yup, I moved the analyzer one to May given the timing of when this would need to be done ('Dart 2.14 issue a deprecation warning').

**Comment by franklinyow** [comment]
Moving CFE to May

**Comment by franklinyow** [comment]
@mraleph Is there anything else we are tracking for April here, now that we moved all the depreciation warning to May?

**Comment by mraleph** [comment]
Moved to May.

(Related: I have migrated 1 internal customer. 2 to go).

**Comment by mraleph** [comment]
Internal customers are fully migrated (the last CL is under review).

Announcement sent.

**Comment by franklinyow** [comment]
@mraleph Are there any other work here for 2.14?  We have tracking issues for the warning messages.  
I can move this out of the May milestone if the 2.14 work is done

**Comment by mraleph** [comment]
No, only changes to the CFE and analyzer are pending. We are all set to go internally.

**Comment by Piero512** [comment]
I have a question about this deprecation. There are some C APIs that are very asyncronous, and recently on dart:ffi there was some work in enabling sending data from arbitrary threads through SendPorts. Is this change going to affect those? https://api.dart.dev/stable/2.13.1/dart-ffi/NativeApi-class.html

**Comment by mraleph** [comment]
@Piero512 no, things you have referenced will continue to work

**Comment by GavinRay97** [comment]
Will it ever be a possibility to use Dart to compile to a dynamic/static lib?

I know this isn't possible at the moment (only a binary), but I'm not sure I get why. I recognize that Dart has a runtime, but the situation is similar for .NET and the solution they adopted was to require the runtime present/loaded and initialized prior to trying to load/call any export symbols from the object.

It would be really cool to have Dart as an option when looking to write native libs.

**Comment by devoncarew** [comment]
I believe we were going to deprecate in 2.14 (shipping soon) and remove in 2.15.

**Comment by a-siva** [comment]
Yes the removal will be in the first beta of 2.15

**Comment by mraleph** [comment]
@GavinRay97 asking off-topic questions on clearly unrelated process issues is not a good strategy to get an answer. We don't have concrete plans to implement what you describe, but we discussed implementing something like that before. The main reason for no implementing it is lack of clear demand and lack of resources. Feel free to file a separate issue with a feature request.

**Comment by mraleph** [comment]
@johnniwinther @srawlins could you make changes to CFE and analyzer to start rejecting `dart-ext:` imports?

**Comment by mraleph** [comment]
@lrhn yeah, I will do few separate changes to replace `native` with an annotation.

**Comment by mraleph** [comment]
VM and CFE has removed support for `dart-ext`. Analyzer work pending. 

Removal of `native` syntax is tracked in https://github.com/dart-lang/sdk/issues/28791

**Comment by devoncarew** [comment]
@mraleph - do you know what the remaining work here is for the stable release? If there is work remaining, perhaps  track using sub-issues and update the first comment in this one to be a checklist / meta issue.
